OSOE-146: Flow 'Additional Styling Part' flyout not activating

dministro opened the pull request:
Fixes GH-76

OSOE-146

dministro committed:
fixing button data attribute

dministro committed:
changing directory structure

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Helpful-Extensions into issue/OSOE-146

dministro committed:
Adding missed documentation.

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/Helpful-Extensions into issue/OSOE-146

dministro committed:
Adding Flow UI test

dministro committed:
Adding basic UI tests for Widgets feature

dministro committed:
Typo

dministro committed:
Comments

sarahelsaig commented on .gitignore:
```diff
@@ -8,3 +8,4 @@ wwwroot/
 node_modules/
 *.user
 .pnpm-debug.log
+.editorconfig
```
Please retain the empty line at the end of the .gitignore.

sarahelsaig commented on Lombiq.HelpfulExtensions.Tests.UI/Constants/XPathSelectors.cs:
```diff
@@ -0,0 +1,19 @@
+public static class XPathSelectors
+{
+    public const string AddWidgetButtonGroup = "//div[contains(@class, 'btn-widget-add-below-wrapper')]"
+        + "/div/div[@class='btn-group']";
```
Always use `contains(@class, '...')`, so it doesn't break if you add an additional item into the class attribute. That's pretty hard to debug later. Same below for other `@className = 'something' paths.

Also unless it's necessary, avoid chains of element names becuase it makes the path unnecessarily rigid. Prefer to only give the necessary structure and use descendant selector (`//`) to skip unnecessary specificity. So this can probably be `+ "//div[contains(@class, 'btn-group')]";`

sarahelsaig commented on Lombiq.HelpfulExtensions.Tests.UI/Constants/XPathSelectors.cs:
```diff
@@ -0,0 +1,19 @@
+        + "/div[@class='btn-group']/div[@class='btn-group']";
+    public const string FlowSettingsButton = $"{FlowSettingsButtonGroup}/button[@title='Settings']";
+    public const string FlowSettingsDropdown = $"{FlowSettingsButtonGroup}/div[contains(@class, 'dropdown-menu')]";
+    public const string CustomClassesInput = $"{FlowSettingsDropdown}/div[@class='form-group']"
+        + "/input[@id='FlowPart-0_ContentItem_CustomClasses']";
```
In HTML the id attribute has to be globally unique (in fact if this isn't fulfilled then `html-validate` will cause the UI test to fail) so there is no point adding any preceding selectors. Also there is a built-in XPath function for selecting ids
```suggestion
    public const string CustomClassesInput = "id('FlowPart-0_ContentItem_CustomClasses')";
```

sarahelsaig commented on Lombiq.HelpfulExtensions.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+
+        var flowSettingsButtonSelector = By.XPath(FlowSettingsButton);
+
+        // Check that, the settings button is exists and visible.
```
```suggestion
        // Check that the settings button exists and is visible.
```

sarahelsaig commented on Lombiq.HelpfulExtensions.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+            .Build()
+            .Perform();
+
+        // Check that, the 'AdditionalStylingPart' view is exists and visible.
```
```suggestion
        // Check that the 'AdditionalStylingPart' view exists and is visible.
```

sarahelsaig commented on Lombiq.HelpfulExtensions.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+
+    public static async Task TestFlowAdditionalStylingPartNotActivatingGh76Async(this UITestContext context)
```
"Gh76"?

dministro replied:
The GitHub issue code. Some open-source projects use issue code in test names, while others don't. I didn't find any suggestions about this. I can remove it.

sarahelsaig replied:
Referencing the issue is a good idea, but this way is not obvious. I'd prefer if you added a `<summary>` above this method that included the issue title and its URL.

sarahelsaig commented on Lombiq.HelpfulExtensions.Tests.UI/Extensions/UITestContextExtensions.cs:
```diff
@@ -0,0 +1,11 @@
+public static class UITestContextExtensions
+{
+    public static Task GoToCreatePageAsync(this UITestContext context) =>
+        context.GoToRelativeUrlAsync("/Admin/Contents/ContentTypes/Page/Create", onlyIfNotAlreadyThere: false);
```
There is already a `Lombiq.Tests.UI.Extensions.OrchardCoreDashboardUITestContextExtensions.CreateNewContentItemAsync` extension method.

sarahelsaig commented on Lombiq.HelpfulExtensions.Tests.UI/Lombiq.HelpfulExtensions.Tests.UI.csproj:
```diff
@@ -0,0 +1,48 @@
+    <PackageIcon>NuGetIcon.png</PackageIcon>
+    <PackageTags>OrchardCore;Lombiq;AspNetCore;CodeGeneration;ShapeTracing;Widgets</PackageTags>
+    <RepositoryUrl>https://github.com/Lombiq/Helpful-Extensions</RepositoryUrl>
+    <PackageProjectUrl>https://github.com/Lombiq/Helpful-Extensions</PackageProjectUrl>
```
https://lombiq.atlassian.net/wiki/spaces/DEV/pages/19496967/Creating+a+new+Visual+Studio+project+or+solution#:~:text=for%20sub-projects%20their%20own%20readme

sarahelsaig commented on Readme.md:
```diff
@@ -100,7 +100,9 @@ To add inline styles include:
 
+### Target blank
 
+Turns all external links into `target="_blank"`.
```
```suggestion
Gives all external links the `target="_blank"` attribute.
```

dministro committed:
Update Lombiq.HelpfulExtensions.Tests.UI/Constants/XPathSelectors.cs

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.HelpfulExtensions.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Lombiq.HelpfulExtensions.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Update Readme.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>

dministro committed:
Fixing missing empty line at the end in `.gitignore`
Fixing XPathSelectors
Adding GH issue link to summary
Changing to `Lombiq.Tests.UI.Extensions.OrchardCoreDashboardUITestContextExtensions.CreateNewContentItemAsync`
Fixing readme link in `Lombiq.HelpfulExtensions.Tests.UI.csproj`

dministro committed:
Adding Lombiq.HelpfulExtensions.Tests.UI NuGet shield to Readme.md

dministro committed:
Changing Lombiq.Tests.UI version to 3.3.1-alpha.osoe-146.0

sarahelsaig commented on Readme.md:
```diff
@@ -3,6 +3,7 @@
 
 
 [![Lombiq.HelpfulExtensions NuGet](https://img.shields.io/nuget/v/Lombiq.HelpfulExtensions?label=Lombiq.HelpfulExtensions)](https://www.nuget.org/packages/Lombiq.HelpfulExtensions/)
+[![Lombiq.HelpfulExtensions NuGet](https://img.shields.io/nuget/v/Lombiq.HelpfulExtensions.Tests.UI?label=Lombiq.HelpfulExtensions.Tests.UI)](https://www.nuget.org/packages/Lombiq.HelpfulExtensions/)
```
```suggestion
[![Lombiq.HelpfulExtensions UI Test Extensions NuGet](https://img.shields.io/nuget/v/Lombiq.HelpfulExtensions.Tests.UI?label=Lombiq.HelpfulExtensions.Tests.UI)](https://www.nuget.org/packages/Lombiq.HelpfulExtensions.Tests.UI/)
```

dministro committed:
Update Readme.md

Co-authored-by: Dávid El-Saig <david.el-saig@lombiq.com>
